core/shared/src/main/scala/org/http4s/Message.scala

```diff
@@ -465,6 +465,22 @@ final class Request[+F[_]] private (
   def isIdempotent: Boolean =
     method.isIdempotent || headers.contains[`Idempotency-Key`]
 
+  /**
+   * Checks whether the [[Entity]] of the [[Message]] is pure in the sense of
+   * indicating that the entity body can be re-run without side effects.
+   *
+   * @return the [[Boolean]] value:
+   *         'true' if entity is empty or already loaded into memory;
+   *         'false' if entity is streamed.
+   */
+  def isPureEntity: Boolean =
+    entity match {
+      case Entity.Empty | Entity.Strict(_) =>
+        true
+      case Entity.Default(_, _) =>
+        false
+    }
```

Comment:
Note that a `Stream`ing entity can also be pure, it's just harder to check this :) something to think about in method naming and/or scaladoc e.g. `isKnownPureEntity` (as in Scala's `knownSize` methods).

Reply:
Btw, should this be `hasPureEntity`, since it's defined on `Message`?

Reply:
If we do some archeological work, in the past http4s contained something you are talking about https://github.com/http4s/http4s/pull/714/files#diff-adcd6450231fbb969572de7bdd97e36c7dd50fe3e577135b505c3a7e90320077R39, but it was removed. But yeah, you're right that we need to rename this method.

Reply:
Renamed to `hasPureEntity`. Let it sound more native.

Reply:
I don't think there's any way to determine whether a stream is pure or not at runtime.  The algebra isn't introspectable like it was in antiquity.  Something like "hasKnownPureEntity" works for me.

Reply:
My two cents about the 'known' suffix - it gives not so many insights. Users still have to check scaladoc/impl to get what it actually means. But If you insist, I'm not that much against it.